implement new syntax for type declaration

tatchi opened the pull request:
This is my attempt at implementing the new syntax as proposed in https://github.com/ocaml-ppx/ppx_import/issues/62.

It only changes the `type declaration` and not the `module type declaration` as the latter is already defined as a regular context-free rule.  However, and for consistency purposes, I think that it would make sense to also change the syntax of the `module type declaration`. So

`module type Hashable = [%import: (module Hashtbl.HashedType)]`

would become

`[%%import: module type Hashable = Hashtbl.HashedType]` or `module type%import Hashable = Hashtbl.HashedType`

I've some local changes that partially address that but if you agree, I would prefer to address that in a separate PR as it seems trickier/requires more changes to the actual implementation.

tatchi commented on src_test/ppx_deriving/test_intf.ml:
```diff
@@ -1 +1 @@
-type a = [%import: Stuff.a]
+[%%import: type a = Stuff.a]
```
That syntax `module%type a = Stuff.a` should work too but is currently not covered by any tests. What would be the best way to add tests for it? Should we duplicate every test or just convert one/some of them to that syntax?

pitag-ha replied:
My opinion on that is the following: making sure that either `[%%import: type a = Stuff.a]` or `module type%import a = Stuff.a` does the right thing is `ppx_import`'s job. Making sure that the two syntaxes `[%%import: type a = Stuff.a]` and `module type%import a = Stuff.a` do the same thing is the compiler's job. So here, I'd say it's enough to test one of the two syntaxes and to assume that the other one does the same.

However, it might be nice to add one single test of the other syntax for two reasons: that way you notice if something changes in the compiler and people who use the tests to look up how the library works can see the alternative syntax (although, tbh, I'd just add it to the README).

Of course, that decision is up to the `ppx_import` maintainers though!

pitag-ha commented:
Thanks a lot for this PR @tatchi! From a `ppxlib` perspective and, as far as I can tell, also from a `ppx_import` perspective, the new syntax is a great improvement!

@ejgallego, will you have time to review this? There's no rush! I'm just checking in case you were expecting me to review or similar.

ejgallego commented:
I can review this, but of course any help would be very welcome!

ejgallego commented:
While I get cycles to look at this, adding a CHANGES entry would help too.

tatchi commented:
> While I get cycles to look at this, adding a CHANGES entry would help @tatchi

I added a change entry. We should probably also adapt the examples in the readme. What syntax do you want to emphasize? `[%%import ... ]` or `type%import ... ` 


`

ejgallego commented:
What would be the plan for this, should we start a 2.0 branch?

If so, should we allow ppx_import 1.x to be coinstallable with 2.x ?

ejgallego commented:
Thanks @tatchi , no idea either on what would be the best path for the 2.0 branch, I cc @gasche and @kit-ty-kate in case they'd like to weight in.

pitag-ha commented:
Sorry for the very long time without answering!

> We should probably also adapt the examples in the readme.

Yes, I agree.

> What syntax do you want to emphasize? [%%import ... ] or type%import ... ?

My opinion on this is that `type%import` is the nicer syntax. However, of course, also this is up to the maintainers!

pitag-ha commented:
> should we allow ppx_import 1.x to be coinstallable with 2.x ?

@ejgallego, that's a good question, and I'm probably not the right one to answer this. My naive way of seeing this is that it's both simpler and better if they're not co-installable. Better because that way the ecosystem is likely going to switch faster to 2.x once the first packages have started doing so.

panglesd commented on src/ppx_import.ml:
```diff
@@ -369,52 +369,48 @@ let type_declaration ~tool_name ~input_name (type_decl : Ppxlib.type_declaration
+            let env = Lazy.force lazy_env in
+            match lid with
+            | Lapply _ ->
+              Location.raise_errorf ~loc
+                "[%%import] cannot import a functor application %s"
+                (string_of_lid lid)
```
Again, this comment is not really related to the PR, but it is a bad practice to raise errors with ppxlib, as it prevents other rewriters to work and more importantly it prevents merlin to work...
Instead, errors should be embedded. I have written a section in the ppxlib manual about his [here](https://ocaml-ppx.github.io/ppxlib/ppxlib/manual.html#handling-errors), which includes a guide for [migrating from raising to embedding](https://ocaml-ppx.github.io/ppxlib/ppxlib/manual.html#migrating-from-raising-to-embedding-errors).

I understand that this should most certainly be part of another PR!

panglesd reviewed (approved):
(I'm reviewing this as a ppxlib maintainer)

Looks good!

My comments in code summed up here are:
- A test for `[%%import type t = Location.t]` (without the `:`) should be added
- A suggestion for supporting `type%import t = Location.t and t2 = Location.t2`
- An unrelated suggestion for error reporting!

I also have another remark. @pitag-ha mentioned the improvement in performance, due to not having to do another pass.
Unfortunately, since `ppx_import` has to be run as a `staged_pps`, its pass won't be merged with other ppx (apart from the staged ones)...

Finally, the README should be modified, but I understand this modification cannot be merged until the new version is released (or, in a 2.0 branch). In this README, I think it should be emphasized that the `[@@deriving]` attribute should be put inside the `[%%import]`!

pitag-ha commented:
> I also have another remark. @pitag-ha mentioned the improvement in performance, due to not having to do another pass.
Unfortunately, since ppx_import has to be run as a staged_pps, its pass won't be merged with other ppx (apart from the staged ones)...

Oh, right!

tatchi commented on src/ppx_import.ml:
```diff
@@ -505,18 +501,38 @@ let module_type ~tool_name ~input_name (package_type : Ppxlib.package_type) =
+  Ppxlib.Extension.V3.declare "import" Ppxlib.Extension.Context.signature_item
+    Ppxlib.Ast_pattern.(
+      psig (psig_type __ (__ ^:: nil) ^:: nil)
+      ||| pstr (pstr_type __ (__ ^:: nil) ^:: nil))
```
No specific reason, I think I just missed that case. Thanks for the suggestion 🤗

tatchi commented:
Thanks for the careful review @panglesd 🙏

> * A test for `[%%import type t = Location.t]` (without the `:`) should be added

I adapted one of the tests to this syntax, and also one with `type%import ....`. See https://github.com/ocaml-ppx/ppx_import/pull/68/commits/e51599245f9ce89d7682fdd4268ac7229c7b18bc

> * A suggestion for supporting `type%import t = Location.t and t2 = Location.t2`

Thanks, that definitely makes sense and seems easy to support 😁. I made the changes in https://github.com/ocaml-ppx/ppx_import/pull/68/commits/19b230e68527f2f2984dd3580095f4ac82d0c306

> * An unrelated suggestion for error reporting!

Interesting, I'll read the doc and see if I can avoid raising errors (that will be for another PR)

ejgallego commented:
Thanks for all the work folks, so I guess it is time to create a 1.x branch and start thinking about merging this in master.

I agree that having a 2.0 release with the same package name is simpler, however the opam repos would need a pretty large patch to introduce a `< 2.0` constraint in all of ppx_import rev deps.

Is this the standard operating procedure?

kit-ty-kate commented:
> Is this the standard operating procedure?

yes, it is totally normal and should be merged rather quickly. In this case you can even use `opam admin add-constraint ppx_import<2.0` unconditionally given (if i understand correctly) everything is going to break

ejgallego commented:
Thanks for the feedback @kit-ty-kate !

So I'm happy to go ahead with that then, @tatchi what's the status of this PR from your part?

tatchi commented:
> So I'm happy to go ahead with that then, @tatchi what's the status of this PR from your part?

I should have addressed all the points @panglesd mentioned in https://github.com/ocaml-ppx/ppx_import/pull/68#pullrequestreview-1088712206. Except for the improvement of the errors for which I opened a separate draft PR https://github.com/ocaml-ppx/ppx_import/pull/73

There's also the README that needs to be adapted, but I believe we can do it in a separate PR too.

ejgallego commented:
Great! I think then this is ready to go, I'll do a last review pass and some internal testing ASAP.

ejgallego commented on CHANGES.md:
````diff
@@ -1,3 +1,19 @@
+  (* or *)
+  type%import loc = Location.t
+  ```
+
````
Would it be possible to explain how the other parts of the syntax work? In particular the `with foo := bar` alias.

Maybe it is better just to update the readme in this PR too? Indeed, it could be a good idea to keep the documentation of the code in sync.

ejgallego reviewed (approved):
LGTM, thanks a lot!

I am not sure however if we should merge this PR without an update to the README; personally I'd prefer to keep the doc in sync at all times.

I will proceed to setup the right branches and opam repository constraints , then merge this PR.

tatchi commented on README.md:
```diff
@@ -35,7 +35,7 @@ Syntax
-# type loc = [%import: Location.t];;
+# type%import loc = Location.t;;
 type loc = Location.t = { loc_start : Lexing.position; loc_end : Lexing.position; loc_ghost : bool; }
 # module type Hashable = [%import: (module Hashtbl.HashedType)];;
```
As I mentioned in https://github.com/ocaml-ppx/ppx_import/pull/68#issue-1122253293, that would be nice to change the syntax of module type declaration too.

tatchi replied:
Happy to open a follow-up PR with address that :)

tatchi commented:
> I am not sure however if we should merge this PR without an update to the README; personally I'd prefer to keep the doc in sync at all times.

I updated the readme with the new syntax. Note that the module type declaration syntax hasn't been modified (because it's already written as a context-free rule). 

IMO, it would make sense to update it to `module type%import ...` to stay consistent. Maybe something to keep in mind before releasing a 2.0 ?

ejgallego commented:
> IMO, it would make sense to update it to `module type%import ...` to stay consistent. Maybe something to keep in mind before releasing a 2.0 ?

Yes! If you think it is a good idea to keep track of this, please open an issue and assign the 2.0 milestone to it.

ejgallego commented:
Thanks a lot for all the work, I've created the `v1.x` branch for patches going to the 1.x series.
